core/shared/src/main/scala/scalaz/zio/internal/Executor.scala

```diff
@@ -0,0 +1,203 @@
+// Copyright (C) 2018 - 2019 John A. De Goes. All rights reserved.
+package scalaz.zio.internal
+
+import java.util.concurrent._
+import scala.concurrent.ExecutionContext
+
+/**
+ * An executor is responsible for executing actions. Each action is guaranteed
+ * to begin execution on a fresh stack frame.
+ */
+trait Executor {
+
+  /**
+   * The concurrency level of the executor.
+   */
+  def concurrency: Int
+
+  /**
+   * The capacity of the executor.
+   */
+  def capacity: Int
+
+  /**
+   * The number of tasks remaining to be executed.
+   */
+  def size: Int
+
+  /**
+   * The number of tasks that have been enqueued, over all time.
+   */
+  def enqueuedCount: Long
+
+  /**
+   * The number of tasks that have been dequeued, over all time.
+   */
+  def dequeuedCount: Long
+
+  /**
+   * Submits a task for execution.
+   */
+  def submit(runnable: Runnable): Boolean
+
+  /**
+   * Initiates shutdown of the executor.
+   */
+  def shutdown(): Unit
+
+  /**
+   * Views this `Executor` as a Scala `ExecutionContext`.
+   */
+  lazy val asExecutionContext: ExecutionContext =
+    new ExecutionContext {
+      override def execute(r: Runnable): Unit =
+        if (!submit(r)) throw new RejectedExecutionException("Rejected: " + r.toString)
+
+      override def reportFailure(cause: Throwable): Unit =
+        cause.printStackTrace
+    }
+}
+
+object Executor extends Serializable {
+  sealed abstract class Type extends Product with Serializable
+  object Type extends Serializable {
+
+    /**
+     * An executor optimized for synchronous tasks, which yield
+     * to the runtime infrequently or never.
+     */
+    final case object Synchronous extends Type
+
+    /**
+     * An executor optimized for asynchronous tasks, which yield
+     * frequently to the runtime.
+     */
+    final case object Asynchronous extends Type
+  }
+
+  /**
+   * Creates a new default executor of the specified type.
+   */
+  final def newDefaultExecutor(tpe: Type): Executor = tpe match {
+    case Type.Synchronous =>
+      fromThreadPoolExecutor {
+        val corePoolSize  = Int.MaxValue
+        val maxPoolSize   = Int.MaxValue
+        val keepAliveTime = 1000L
+        val timeUnit      = TimeUnit.MILLISECONDS
+        val workQueue     = new SynchronousQueue[Runnable]()
+        val threadFactory = new NamedThreadFactory("zio-blocking", true)
```

Comment:
At least for me, I have to change this every time, because I need to propagate the thread-local every time. see https://github.com/akka/akka/issues/26056

Reply:
`zio-default-sync`

Reply:
See `FiberLocal`.

Reply:
Yeah, I'll change the name. 👍